File: 202_HappyNumber.py
# 93.68%
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        s = set()
        while n is not 1:
            if n not in s:
                s.add(n)
                temp = 0
                while n:
                    temp = temp + (n % 10) ** 2
                    n /= 10
                n = temp
            else:
                return False
        return True
		
def isHappy(self, n):
    mem = set()
    while n != 1:
        n = sum([int(i) ** 2 for i in str(n)])
        if n in mem:
            return False
        else:
            mem.add(n)
    else:
        return True
                
            

File: 202_HappyNumber.cpp
//11.30%
class Solution {
public:
    bool isHappy(int n) {
        while(n != 1){
			if(table.find(n) == table.end()){
				table[n] = 1;
				int m = 0;
				while(n){
					int temp = n % 10;
					m += (temp * temp);
					n /= 10;
				}
				n = m;
			}else
				return false;
		}
		return true;			
    }
private:
	map<int, int> table; //or unordered_map
};

//11.30%
class Solution {
public:
    bool isHappy(int n) {
        while(n != 1){
			if(table.find(n) == table.end()){
				table.insert(1);
				int m = 0;
				while(n){
					int temp = n % 10;
					m += (temp * temp);
					n /= 10;
				}
				n = m;
			}else
				return false;
		}
		return true;			
    }
private:
	set<int> table; //unordered_set
};